common/src/main/java/io/netty/util/concurrent/Futures.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.concurrent;
+
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.util.Objects;
+import java.util.function.Function;
+
+import static io.netty.util.internal.PromiseNotificationUtil.tryCancel;
+import static io.netty.util.internal.PromiseNotificationUtil.tryFailure;
+
+/**
+ * Combinator operations on {@linkplain Future futures}.
+ * Used for implementing {@link Future#map(Function)} and {@link Future#flatMap(Function)}.
+ *
+ * @implNote The operations themselves are implemented as static inner classes instead of lambdas to aid debugging.
+ */
+final class Futures {
+    private static final InternalLogger LOGGER = InternalLoggerFactory.getInstance(Futures.class);
+    private static final PassThrough<?> PASS_THROUGH_CONSTANT = new PassThrough<Object>();
+
+    /**
+     * @see Future#map(Function)
+     */
+    static <V, R> Future<R> map(Future<V> future, Function<V, R> mapper) {
+        Objects.requireNonNull(future, "The future cannot be null.");
+        Objects.requireNonNull(mapper, "The mapper function cannot be null.");
+        Promise<R> promise = future.executor().newPromise();
+        future.addListener(new Mapper<>(promise, mapper));
```

Comment:
Should we check if the `future` is already done before adding the listener?

Reply:
I thought about it, but then the listener would run immediately instead of through the executor, and that could be a surprise - especially if the current thread is not an event loop thread.

Reply:
let's just always use the listener for now as the promise itself could also do something as part of adding the listener for decoration etc.

Reply:
@normanmaurer You approved the PR, but do you want me to remove the short-circuiting I added? https://github.com/netty/netty/pull/11607/commits/14a9f20fff220db6642d67748b69d9c5aa5c21ef#diff-a7eb5753c6ab028c201ce7cc70593d9ce16f3d386d9f353644c49b4a7d74dca2R52

Reply:
good question. Maybe just leave it as it is and we can revisit later as we have other places as well which does something similar. Maybe someone shouldn't really make any assumptions anyway as long as we ensure the method is called from the same `EventExecutor` thread as the `Future` is assigned too .